ca/external.go

```diff
@@ -0,0 +1,151 @@
+package ca
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/cloudflare/cfssl/api"
+	"github.com/cloudflare/cfssl/signer"
+)
+
+// ErrNoExternalCAURLs is an error used it indicate that an ExternalCA is
+// configured with no URLs to which it can proxy certificate signing requests.
+var ErrNoExternalCAURLs = errors.New("no external CA URLs")
+
+// ExternalCA is able to make certificate signing requests to one of a list
+// remote CFSSL API endpoints.
+type ExternalCA struct {
+	sync.Mutex
+
+	rootCA *RootCA
+	urls   []string
+	client *http.Client
+}
+
+// NewExternalCA creates a new ExternalCA which uses the given tlsConfig to
+// authenticate to any of the given URLS of CFSSL API endpoints.
+func NewExternalCA(rootCA *RootCA, tlsConfig *tls.Config, urls ...string) *ExternalCA {
+	return &ExternalCA{
+		rootCA: rootCA,
+		urls:   urls,
+		client: &http.Client{
+			Transport: &http.Transport{
+				TLSClientConfig: tlsConfig,
+			},
+		},
+	}
+}
+
+// UpdateTLSConfig updates the HTTP Client for this ExternalCA by creating
+// a new client which uses the given tlsConfig.
+func (eca *ExternalCA) UpdateTLSConfig(tlsConfig *tls.Config) {
+	eca.Lock()
+	defer eca.Unlock()
+
+	eca.client = &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+	}
+}
+
+// UpdateURLs updates the list of CSR API endpoints by setting it to the given
+// urls.
+func (eca *ExternalCA) UpdateURLs(urls ...string) {
+	eca.Lock()
+	defer eca.Unlock()
+
+	eca.urls = urls
+}
+
+// CanSign returns whether this ExternalCA is ready to sign, i.e., whether
+// there it as least one remote URL configured.
+func (eca *ExternalCA) CanSign() bool {
```

Comment:
I'm not sure it makes sense to have a `CanSign` function when the ability to sign may change by the time we call `Sign`. How about a flow like this:

```go
// Try signing with an external CA first
cert, err = externalCA.Sign(PrepareCSR(rawCSR, cn, ou, org))

if err == ErrNoExternalCAURLs {
	// No external CA URLs configured. Try signing locally.
	cert, err = rootCA.ParseValidateAndSignCSR(rawCSR, cn, ou, org)
}

if err != nil {
	...
}
```